File: v4/rust/iprange-livedb/src/validation/bitmap.rs
use crate::bitmap_page::{self, Header, HeaderProblem, BRANCH_CHILDREN, LEAF_WORDS, MAX_LEVEL};
use crate::error::{Error, Result};
use crate::mapping::ByteSource;

use super::context::Context;
use super::{ValidationObject, ValidationReason, ValidationSink};

mod query;
mod word_cache;

pub(crate) use query::contains;
use query::require_header as require_query_header;
pub(crate) use word_cache::WordCache;

pub(crate) use crate::bitmap_page::Kind;

trait ValidationKind {
    fn object(self) -> ValidationObject;
}

impl ValidationKind for Kind {
    fn object(self) -> ValidationObject {
        match self {
            Kind::Free => ValidationObject::FreeBitmap,
            Kind::Feed => ValidationObject::FeedUsedBitmap,
            Kind::Membership => ValidationObject::MembershipUsedBitmap,
        }
    }
}

pub(crate) fn validate<S: ValidationSink>(
    context: &mut Context<'_, S>,
    root: u32,
    limit: u64,
    kind: Kind,
) -> Result<u64> {
    if root == 0 {
        return Ok(0);
    }
    if limit == 0 {
        context.emit(
            ValidationReason::BitmapSummaryInvalid,
            kind.object(),
            Some(root),
            None,
            None,
        )?;
        return Ok(0);
    }

    let required = required_level(limit)?;
    let mut path = [0; MAX_LEVEL as usize + 1];
    let result = validate_node(context, root, required, 0, limit, kind, &mut path, 0)?;
    Ok(result.map_or(0, |result| result.set_bits))
}

#[allow(clippy::too_many_arguments)]
fn validate_node<S: ValidationSink>(
    context: &mut Context<'_, S>,
    page_number: u32,
    expected_level: u16,
    base: u64,
    limit: u64,
    kind: Kind,
    path: &mut [u32; MAX_LEVEL as usize + 1],
    depth: usize,
) -> Result<Option<NodeResult>> {
    let Some(slot) = path.get_mut(depth) else {
        context.emit(
            ValidationReason::TreeLevelInvalid,
            kind.object(),
            Some(page_number),
            None,
            None,
        )?;
        return Ok(None);
    };
    *slot = page_number;
    let Some(page) = context.read_graph_page(page_number, kind.object(), &path[..depth])? else {
        return Ok(None);
    };
    let Some(header) = parse_header(context, page_number, page, expected_level, kind)? else {
        return Ok(None);
    };
    if header.level == 0 {
        validate_leaf(context, page_number, page, base, limit, kind, header)
    } else {
        validate_branch(
            context,
            page_number,
            page,
            base,
            limit,
            kind,
            header,
            path,
            depth,
        )
    }
}

fn parse_header<S: ValidationSink, P: ByteSource>(
    context: &mut Context<'_, S>,
    page_number: u32,
    page: P,
    expected_level: u16,
    kind: Kind,
) -> Result<Option<Header>> {
    let header =
        match bitmap_page::checked_header(page, context.meta.txn_id, kind, Some(expected_level)) {
            Ok(header) => header,
            Err(problem) => {
                emit_header_problem(context, page_number, kind, problem)?;
                return Ok(None);
            }
        };
    if !bitmap_page::reserved_zero(page, header.level) {
        context.emit(
            ValidationReason::PageReservedNonzero,
            kind.object(),
            Some(page_number),
            None,
            None,
        )?;
    }
    Ok(Some(header))
}

fn emit_header_problem<S: ValidationSink>(
    context: &mut Context<'_, S>,
    page_number: u32,
    kind: Kind,
    problem: HeaderProblem,
) -> Result<()> {
    let reason = match problem {
        HeaderProblem::Header => ValidationReason::PageHeaderInvalid,
        HeaderProblem::Born => ValidationReason::PageBornTxnInvalid,
        HeaderProblem::Level => ValidationReason::TreeLevelInvalid,
        HeaderProblem::Type => ValidationReason::PageTypeMismatch,
    };
    context.emit(reason, kind.object(), Some(page_number), None, None)
}

fn validate_leaf<S: ValidationSink, P: ByteSource>(
    context: &mut Context<'_, S>,
    page_number: u32,
    page: P,
    base: u64,
    limit: u64,
    kind: Kind,
    header: Header,
) -> Result<Option<NodeResult>> {
    let mut totals = LeafTotals::default();
    for index in 0..LEAF_WORDS {
        let word = bitmap_page::leaf_word(page, index)?;
        let (word_base, valid, valid_mask) =
            validate_leaf_word(context, page_number, base, limit, kind, index, word)?;
        totals.add(context, word_base, word, valid, valid_mask, kind)?;
    }
    if totals.nonzero_words != header.item_count {
        context.emit(
            ValidationReason::PageHeaderInvalid,
            kind.object(),
            Some(page_number),
            None,
            None,
        )?;
    }
    Ok(Some(totals.result()))
}

#[derive(Default)]
struct LeafTotals {
    nonzero_words: usize,
    set_bits: u64,
    has_candidate: bool,
}

impl LeafTotals {
    fn add<S: ValidationSink>(
        &mut self,
        context: &mut Context<'_, S>,
        word_base: u64,
        word: u64,
        valid: u64,
        valid_mask: u64,
        kind: Kind,
    ) -> Result<()> {
        self.nonzero_words += usize::from(word != 0);
        self.set_bits = self
            .set_bits
            .checked_add(u64::from(valid.count_ones()))
            .ok_or(Error::ArithmeticOverflow("validation bitmap bit count"))?;
        match kind {
            Kind::Free => mark_free_bits(context, word_base, valid)?,
            Kind::Feed | Kind::Membership => {
                self.has_candidate |= (!valid) & valid_mask != 0;
            }
        }
        Ok(())
    }

    fn result(self) -> NodeResult {
        NodeResult {
            set_bits: self.set_bits,
            has_one: self.set_bits != 0,
            has_candidate: self.has_candidate,
        }
    }
}

#[allow(clippy::too_many_arguments)]
fn validate_leaf_word<S: ValidationSink>(
    context: &mut Context<'_, S>,
    page_number: u32,
    base: u64,
    limit: u64,
    kind: Kind,
    index: usize,
    word: u64,
) -> Result<(u64, u64, u64)> {
    let word_base = base
        .checked_add((index * 64) as u64)
        .ok_or(Error::ArithmeticOverflow("validation bitmap word offset"))?;
    let valid_mask = in_range_mask(word_base, limit, kind);
    if word & !valid_mask != 0 {
        context.emit(
            ValidationReason::BitmapSummaryInvalid,
            kind.object(),
            Some(page_number),
            None,
            None,
        )?;
    }
    Ok((word_base, word & valid_mask, valid_mask))
}

#[allow(clippy::too_many_arguments)]
fn validate_branch<S: ValidationSink, P: ByteSource>(
    context: &mut Context<'_, S>,
    page_number: u32,
    page: P,
    base: u64,
    limit: u64,
    kind: Kind,
    header: Header,
    path: &mut [u32; MAX_LEVEL as usize + 1],
    depth: usize,
) -> Result<Option<NodeResult>> {
    let child_span = coverage(header.level - 1)?;
    let mut totals = BranchTotals::default();
    for index in 0..BRANCH_CHILDREN {
        let child = bitmap_page::branch_child(page, index)?;
        totals.child_count += usize::from(child != 0);
        let result = validate_branch_child(
            context,
            child,
            index,
            base,
            child_span,
            limit,
            kind,
            header.level,
            path,
            depth,
        )?;
        let Some(result) = result else {
            continue;
        };
        totals.add(context, page_number, page, index, kind, result)?;
    }
    if totals.child_count != header.item_count {
        context.emit(
            ValidationReason::PageHeaderInvalid,
            kind.object(),
            Some(page_number),
            None,
            None,
        )?;
    }
    Ok(Some(totals.result()))
}

#[derive(Default)]
struct BranchTotals {
    child_count: usize,
    set_bits: u64,
    has_one: bool,
    has_candidate: bool,
}

impl BranchTotals {
    fn add<S: ValidationSink, P: ByteSource>(
        &mut self,
        context: &mut Context<'_, S>,
        page_number: u32,
        page: P,
        index: usize,
        kind: Kind,
        result: NodeResult,
    ) -> Result<()> {
        let expected = match kind {
            Kind::Free => result.has_one,
            Kind::Feed | Kind::Membership => result.has_candidate,
        };
        if bitmap_page::summary_bit(page, index)? != expected {
            context.emit(
                ValidationReason::BitmapSummaryInvalid,
                kind.object(),
                Some(page_number),
                None,
                None,
            )?;
        }
        self.set_bits = self
            .set_bits
            .checked_add(result.set_bits)
            .ok_or(Error::ArithmeticOverflow("validation bitmap bit count"))?;
        self.has_one |= result.has_one;
        self.has_candidate |= result.has_candidate;
        Ok(())
    }

    fn result(self) -> NodeResult {
        NodeResult {
            set_bits: self.set_bits,
            has_one: self.has_one,
            has_candidate: self.has_candidate,
        }
    }
}

#[allow(clippy::too_many_arguments)]
fn validate_branch_child<S: ValidationSink>(
    context: &mut Context<'_, S>,
    child: u32,
    index: usize,
    base: u64,
    child_span: u64,
    limit: u64,
    kind: Kind,
    level: u16,
    path: &mut [u32; MAX_LEVEL as usize + 1],
    depth: usize,
) -> Result<Option<NodeResult>> {
    let child_base = bitmap_child_base(base, child_span, index)?;
    if child == 0 {
        return Ok(Some(absent_result(child_base, child_span, limit, kind)));
    }
    validate_node(
        context,
        child,
        level - 1,
        child_base,
        limit,
        kind,
        path,
        depth + 1,
    )
}

fn bitmap_child_base(base: u64, span: u64, index: usize) -> Result<u64> {
    let offset = span
        .checked_mul(index as u64)
        .ok_or(Error::ArithmeticOverflow("validation bitmap child offset"))?;
    base.checked_add(offset)
        .ok_or(Error::ArithmeticOverflow("validation bitmap child offset"))
}

fn absent_result(base: u64, span: u64, limit: u64, kind: Kind) -> NodeResult {
    let end = base.saturating_add(span).min(limit);
    NodeResult {
        set_bits: 0,
        has_one: false,
        has_candidate: base.max(kind.first_candidate()) < end,
    }
}

fn mark_free_bits<S: ValidationSink>(
    context: &mut Context<'_, S>,
    word_base: u64,
    mut word: u64,
) -> Result<()> {
    while word != 0 {
        let bit = u64::from(word.trailing_zeros());
        let page = word_base
            .checked_add(bit)
            .ok_or(Error::ArithmeticOverflow("validation free-page number"))?;
        context.mark_allocated(page as u32, ValidationObject::FreeBitmap)?;
        word &= word - 1;
    }
    Ok(())
}

fn in_range_mask(base: u64, limit: u64, kind: Kind) -> u64 {
    if base >= limit {
        return 0;
    }
    let valid = (limit - base).min(64);
    let mut mask = if valid == 64 {
        u64::MAX
    } else {
        (1u64 << valid) - 1
    };
    let first = kind.first_candidate();
    if base < first {
        let excluded = (first - base).min(64);
        mask &= u64::MAX.checked_shl(excluded as u32).unwrap_or(0);
    }
    if kind == Kind::Free && base == 0 {
        mask &= !3;
    }
    mask
}

fn coverage(level: u16) -> Result<u64> {
    bitmap_page::coverage(level)
}

fn required_level(limit: u64) -> Result<u16> {
    bitmap_page::required_level(limit)
}

#[derive(Clone, Copy)]
struct NodeResult {
    set_bits: u64,
    has_one: bool,
    has_candidate: bool,
}

File: v4/go/internal/validation/bitmap.go
package validation

// Bitmap validation (Rust validation/bitmap.rs): the hierarchical bitmap
// walk with the exact reason mapping (header problems, out-of-range
// summary bits, summary-bit and item-count disagreements), the bit
// accounting, and the free-kind allocation marking. The catalog
// used-bitmap arm composes this walk; the free/membership/structure
// arms arrive with their slices.

import (
	"math/bits"

	"github.com/firehol/iprange/v4/go/internal/bitmap"
	"github.com/firehol/iprange/v4/go/internal/format"
)

// bitmapObject maps one bitmap kind to its validation object (Rust
// ValidationKind::object).
func bitmapObject(kind bitmap.Kind) ValidationObject {
	switch kind {
	case bitmap.KindFree:
		return ObjectFreeBitmap
	case bitmap.KindFeed:
		return ObjectFeedUsedBitmap
	case bitmap.KindMembership:
		return ObjectMembershipUsedBitmap
	default:
		return ObjectStructureUsedBitmap
	}
}

// validateBitmap walks one bitmap root and returns the count of set
// bits (Rust bitmap::validate): a zero root is empty, a zero limit with
// a root is the BitmapSummaryInvalid class, and the walk follows the
// graph claims of the kind's object.
func validateBitmap(ctx *context, root uint32, limit uint64, kind bitmap.Kind) (uint64, error) {
	if root == 0 {
		return 0, nil
	}
	if limit == 0 {
		if err := ctx.emit(ReasonBitmapSummaryInvalid, bitmapObject(kind), &root, nil, nil); err != nil {
			return 0, err
		}
		return 0, nil
	}
	required, err := bitmap.RequiredLevel(limit)
	if err != nil {
		return 0, formatError(err)
	}
	var path [bitmap.MaxLevel + 1]uint32
	result, _, err := validateBitmapNode(ctx, root, required, 0, limit, kind, &path, 0)
	if err != nil {
		return 0, err
	}
	if result == nil {
		return 0, nil
	}
	return result.setBits, nil
}

// bitmapNodeResult mirrors Rust NodeResult: the counted set bits and
// the has-one/has-candidate summary contributions.
type bitmapNodeResult struct {
	setBits      uint64
	hasOne       bool
	hasCandidate bool
}

// validateBitmapNode visits one bitmap node (Rust validate_node: the
// path slot, the graph claim, the classified header, and the
// leaf/branch split).
func validateBitmapNode(ctx *context, pageNumber uint32, expectedLevel uint16, base, limit uint64, kind bitmap.Kind, path *[bitmap.MaxLevel + 1]uint32, depth int) (*bitmapNodeResult, uint16, error) {
	if depth >= len(path) {
		if err := ctx.emit(ReasonTreeLevelInvalid, bitmapObject(kind), &pageNumber, nil, nil); err != nil {
			return nil, 0, err
		}
		return nil, 0, nil
	}
	path[depth] = pageNumber
	page, err := ctx.readGraphPage(pageNumber, bitmapObject(kind), path[:depth])
	if err != nil || page == nil {
		return nil, 0, err
	}
	header, problem := bitmap.CheckedHeader(page, ctx.meta.TxnID, kind, &expectedLevel)
	if problem != bitmap.HeaderProblemNone {
		if err := ctx.emit(bitmapHeaderProblemReason(problem), bitmapObject(kind), &pageNumber, nil, nil); err != nil {
			return nil, 0, err
		}
		return nil, 0, nil
	}
	if !bitmap.ReservedZero(page, header.Level) {
		if err := ctx.emit(ReasonPageReservedNonzero, bitmapObject(kind), &pageNumber, nil, nil); err != nil {
			return nil, 0, err
		}
	}
	if header.Level == 0 {
		result, err := validateBitmapLeaf(ctx, pageNumber, page, base, limit, kind, header)
		return result, header.Level, err
	}
	result, err := validateBitmapBranch(ctx, pageNumber, page, base, limit, kind, header, path, depth)
	return result, header.Level, err
}

// bitmapHeaderProblemReason maps one bitmap header problem to its reason
// class (Rust emit_header_problem).
func bitmapHeaderProblemReason(problem bitmap.HeaderProblem) ValidationReason {
	switch problem {
	case bitmap.HeaderProblemBorn:
		return ReasonPageBornTxnInvalid
	case bitmap.HeaderProblemLevel:
		return ReasonTreeLevelInvalid
	case bitmap.HeaderProblemType:
		return ReasonPageTypeMismatch
	default:
		return ReasonPageHeaderInvalid
	}
}

// validateBitmapLeaf validates the 500 words of one bitmap leaf (Rust
// validate_leaf): every out-of-range bit is the BitmapSummaryInvalid
// class, the nonzero-word count must equal the item count, and the
// free kind marks every set bit in the allocation partition.
func validateBitmapLeaf(ctx *context, pageNumber uint32, page []byte, base, limit uint64, kind bitmap.Kind, header bitmap.Header) (*bitmapNodeResult, error) {
	totals := bitmapLeafTotals{}
	for index := 0; index < bitmap.LeafWords; index++ {
		word, err := bitmap.LeafWord(page, index)
		if err != nil {
			return nil, formatError(err)
		}
		wordBase, valid, validMask, err := validateBitmapWord(ctx, pageNumber, base, limit, kind, index, word)
		if err != nil {
			return nil, err
		}
		if err := totals.add(ctx, wordBase, word, valid, validMask, kind); err != nil {
			return nil, err
		}
	}
	if totals.nonzeroWords != uint64(header.ItemCount) {
		if err := ctx.emit(ReasonPageHeaderInvalid, bitmapObject(kind), &pageNumber, nil, nil); err != nil {
			return nil, err
		}
	}
	result := totals.result()
	return result, nil
}

// bitmapLeafTotals mirrors Rust LeafTotals.
type bitmapLeafTotals struct {
	nonzeroWords uint64
	setBits      uint64
	hasCandidate bool
}

func (t *bitmapLeafTotals) add(ctx *context, wordBase uint64, word, valid, validMask uint64, kind bitmap.Kind) error {
	if word != 0 {
		t.nonzeroWords++
	}
	ones := uint64(bits.OnesCount64(valid))
	if t.setBits > ^uint64(0)-ones {
		return &format.Error{Code: format.CodeArithmeticOverflow, Detail: "validation bitmap bit count"}
	}
	t.setBits += ones
	switch kind {
	case bitmap.KindFree:
		return markFreeBits(ctx, wordBase, valid)
	case bitmap.KindFeed, bitmap.KindMembership, bitmap.KindStructure:
		if (^valid)&validMask != 0 {
			t.hasCandidate = true
		}
	}
	return nil
}

func (t *bitmapLeafTotals) result() *bitmapNodeResult {
	return &bitmapNodeResult{setBits: t.setBits, hasOne: t.setBits != 0, hasCandidate: t.hasCandidate}
}

// validateBitmapWord validates one 64-bit word against the limit window
// (Rust validate_leaf_word): bits outside the valid range are the
// BitmapSummaryInvalid class and the returned valid word is bit-masked.
func validateBitmapWord(ctx *context, pageNumber uint32, base, limit uint64, kind bitmap.Kind, index int, word uint64) (uint64, uint64, uint64, error) {
	wordBase := base + uint64(index*64)
	if wordBase < base {
		return 0, 0, 0, &format.Error{Code: format.CodeArithmeticOverflow, Detail: "validation bitmap word offset"}
	}
	validMask := inRangeMask(wordBase, limit, kind)
	if word&^validMask != 0 {
		if err := ctx.emit(ReasonBitmapSummaryInvalid, bitmapObject(kind), &pageNumber, nil, nil); err != nil {
			return 0, 0, 0, err
		}
	}
	return wordBase, word & validMask, validMask, nil
}

// inRangeMask mirrors Rust in_range_mask: the valid bit window of one
// word (limit, the kind's first candidate, and the free bitmap's
// reserved page-0 pair).
func inRangeMask(base, limit uint64, kind bitmap.Kind) uint64 {
	if base >= limit {
		return 0
	}
	valid := limit - base
	if valid > 64 {
		valid = 64
	}
	var mask uint64
	if valid == 64 {
		mask = ^uint64(0)
	} else {
		mask = (1 << valid) - 1
	}
	first := kind.FirstCandidate()
	if base < first {
		excluded := first - base
		if excluded > 64 {
			excluded = 64
		}
		mask &= ^uint64(0) << excluded
	}
	if kind == bitmap.KindFree && base == 0 {
		mask &^= uint64(3)
	}
	return mask
}

// markFreeBits marks every set bit of one free word in the allocation
// partition (Rust mark_free_bits).
func markFreeBits(ctx *context, wordBase uint64, w uint64) error {
	for w != 0 {
		bit := uint64(bits.TrailingZeros64(w))
		page := wordBase + bit
		if page < wordBase {
			return &format.Error{Code: format.CodeArithmeticOverflow, Detail: "validation free-page number"}
		}
		if err := ctx.markAllocated(uint32(page), ObjectFreeBitmap); err != nil {
			return err
		}
		w &= w - 1
	}
	return nil
}

// validateBitmapBranch validates the 256 children of one bitmap branch
// (Rust validate_branch): the nonzero-child count must equal the item
// count, every present child's summary bit must equal the expected
// contribution of its subtree, and absent children of a non-free kind
// still contribute their candidate span.
func validateBitmapBranch(ctx *context, pageNumber uint32, page []byte, base, limit uint64, kind bitmap.Kind, header bitmap.Header, path *[bitmap.MaxLevel + 1]uint32, depth int) (*bitmapNodeResult, error) {
	childSpan, err := bitmap.Coverage(header.Level - 1)
	if err != nil {
		return nil, formatError(err)
	}
	totals := bitmapBranchTotals{}
	for index := 0; index < bitmap.BranchChildren; index++ {
		child, err := bitmap.BranchChild(page, index)
		if err != nil {
			return nil, formatError(err)
		}
		if child != 0 {
			totals.childCount++
		}
		var result *bitmapNodeResult
		if child == 0 {
			childBase, err := bitmapChildBase(base, childSpan, index)
			if err != nil {
				return nil, err
			}
			result = absentBitmapResult(childBase, childSpan, limit, kind)
		} else {
			childBase, err := bitmapChildBase(base, childSpan, index)
			if err != nil {
				return nil, err
			}
			result, _, err = validateBitmapNode(ctx, child, header.Level-1, childBase, limit, kind, path, depth+1)
			if err != nil {
				return nil, err
			}
			if result == nil {
				continue
			}
		}
		if err := totals.add(ctx, pageNumber, page, index, kind, result); err != nil {
			return nil, err
		}
	}
	if totals.childCount != uint64(header.ItemCount) {
		if err := ctx.emit(ReasonPageHeaderInvalid, bitmapObject(kind), &pageNumber, nil, nil); err != nil {
			return nil, err
		}
	}
	result := totals.result()
	return result, nil
}

// bitmapBranchTotals mirrors Rust BranchTotals.
type bitmapBranchTotals struct {
	childCount   uint64
	setBits      uint64
	hasOne       bool
	hasCandidate bool
}

func (t *bitmapBranchTotals) add(ctx *context, pageNumber uint32, page []byte, index int, kind bitmap.Kind, result *bitmapNodeResult) error {
	expected := result.hasOne
	if kind != bitmap.KindFree {
		expected = result.hasCandidate
	}
	summary, err := bitmap.SummaryBit(page, index)
	if err != nil {
		return formatError(err)
	}
	if summary != expected {
		if err := ctx.emit(ReasonBitmapSummaryInvalid, bitmapObject(kind), &pageNumber, nil, nil); err != nil {
			return err
		}
	}
	if t.setBits > ^uint64(0)-result.setBits {
		return &format.Error{Code: format.CodeArithmeticOverflow, Detail: "validation bitmap bit count"}
	}
	t.setBits += result.setBits
	t.hasOne = t.hasOne || result.hasOne
	t.hasCandidate = t.hasCandidate || result.hasCandidate
	return nil
}

func (t *bitmapBranchTotals) result() *bitmapNodeResult {
	return &bitmapNodeResult{setBits: t.setBits, hasOne: t.hasOne, hasCandidate: t.hasCandidate}
}

// bitmapChildBase computes the base bit of one child span (Rust
// bitmap_child_base).
func bitmapChildBase(base, span uint64, index int) (uint64, error) {
	offset := span * uint64(index)
	if span != 0 && offset/span != uint64(index) {
		return 0, &format.Error{Code: format.CodeArithmeticOverflow, Detail: "validation bitmap child offset"}
	}
	childBase := base + offset
	if childBase < base {
		return 0, &format.Error{Code: format.CodeArithmeticOverflow, Detail: "validation bitmap child offset"}
	}
	return childBase, nil
}

// absentBitmapResult mirrors Rust absent_result: an absent child of a
// non-free kind still contributes its candidate span.
func absentBitmapResult(base, span, limit uint64, kind bitmap.Kind) *bitmapNodeResult {
	end := base + span
	if end < base || end > limit {
		end = limit
	}
	candidate := false
	if kind != bitmap.KindFree {
		first := kind.FirstCandidate()
		low := base
		if low < first {
			low = first
		}
		candidate = low < end
	}
	return &bitmapNodeResult{setBits: 0, hasOne: false, hasCandidate: candidate}
}

// validateFreeBitmap runs the free bitmap validator over the whole
// committed generation (Rust validate_selected's bitmap::validate with
// Kind::Free and the page count as the bit limit).
func validateFreeBitmap(ctx *context) error {
	_, err := validateBitmap(ctx, ctx.meta.FreeBitmapRoot, ctx.meta.PageCount, bitmap.KindFree)
	return err
}
